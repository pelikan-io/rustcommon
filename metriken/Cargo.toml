[package]
name = "metriken"
version = "0.4.0"
edition = "2021"
authors = [
	"Brian Martin <brian@iop.systems>",
	"Sean Lynch <sean@iop.systems>",
]
license = "Apache-2.0"
description = "A fast and lightweight metrics library"
homepage = "https://github.com/pelikan-io/rustcommon"
repository = "https://github.com/pelikan-io/rustcommon"

# we use this to prevent multiple versions of metriken from being used in the
# same project
links = "metriken"

[dependencies]
metriken-derive = { version = "=0.4.0", path = "./derive" }
histogram = { version = "0.9.0", path = "../histogram" }

linkme = "0.3.3"
once_cell = "1.14.0"
parking_lot = "0.12.1"

phf = { version = "0.11.2", features = ["macros"] }

[dev-dependencies]
trybuild = "1.0"
