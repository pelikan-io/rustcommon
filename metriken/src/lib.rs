// Copyright 2021 Twitter, Inc.
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

//! Easily registered distributed metrics.
//!
//! More docs todo...
//!
//! # Creating a Metric
//! Registering a metric is straightforward. All that's needed is to declare a
//! static within the [`metric`] macro. By default, the metric will have the
//! name of the path to the static variable you used to declare it but this can
//! be overridden by passing the `name` parameter to the macro.
//!
//! ```
//! # // This should remain in sync with the example below.
//! use metriken::*;
//! /// A counter metric named "<crate name>::COUNTER_A"
//! #[metric]
//! static COUNTER_A: Counter = Counter::new();
//!
//! /// A counter metric named "my.metric.name"
//! #[metric(name = "my.metric.name")]
//! static COUNTER_B: Counter = Counter::new();
//! #
//! # let metrics = metrics();
//! # // Metrics may be in any arbitrary order
//! # let mut names: Vec<_> = metrics.iter().map(|metric| metric.name()).collect();
//! # names.sort();
//! #
//! # assert_eq!(names.len(), 2);
//! # assert_eq!(names[0], "COUNTER_A");
//! # assert_eq!(names[1], "my.metric.name");
//! ```
//!
//! # Accessing Metrics
//! All metrics registered via the [`metric`] macro can be accessed by calling
//! the [`metrics`] function. This will return an instance of the [`Metric`]
//! struct which allows you to access all staticly and dynamically registered
//! metrics.
//!
//! Suppose we have the metrics declared in the example above.
//! ```
//! # // This should remain in sync with the example above.
//! # use metriken::*;
//! # /// A counter metric named "<crate name>::COUNTER_A"
//! # #[metric]
//! # static COUNTER_A: Counter = Counter::new();
//! #
//! # /// A counter metric named "my.metric.name"
//! # #[metric(name = "my.metric.name")]
//! # static COUNTER_B: Counter = Counter::new();
//! #
//! let metrics = metrics();
//! // Metrics may be in any arbitrary order
//! let mut names: Vec<_> = metrics.iter().map(|metric| metric.name()).collect();
//! names.sort();
//!
//! assert_eq!(names.len(), 2);
//! assert_eq!(names[0], "COUNTER_A");
//! assert_eq!(names[1], "my.metric.name");
//! ```
//!
//! # How it Works
//! Behind the scenes, this crate uses the [`linkme`] crate to create a
//! distributed slice containing a [`MetricEntry`] instance for each metric that
//! is registered via the [`metric`] attribute.

mod counter;
mod gauge;
pub mod histogram;
mod lazy;

extern crate self as metriken;

#[doc(inline)]
pub use metriken_core::{
    default_formatter, dynmetrics, metrics, DynMetricsIter, Format, Metadata, MetadataIter, Metric,
    MetricEntry, Metrics, MetricsIter, Value,
};

pub use crate::counter::Counter;
#[doc(inline)]
pub use crate::dynmetrics::{DynBoxedMetric, DynPinnedMetric, MetricBuilder};
pub use crate::gauge::Gauge;
pub use crate::histogram::{AtomicHistogram, RwLockHistogram};
pub use crate::lazy::Lazy;

pub use metriken_derive::metric;

/// A counter holds a unsigned 64bit monotonically non-decreasing value. The
/// counter behavior is to wrap on overflow.
///
/// Common examples are the number of operations (requests, reads, ...) or
/// errors.
///
/// Unlike a standard `Counter`, a `LazyCounter` will not report a value unless
/// it has been initialized by writing to at least once. This is useful for when
/// you want to declare metrics statically, but only report metrics that are
/// being used.
pub type LazyCounter = Lazy<Counter>;

/// A gauge holds a signed 64-bit value and is used to represent metrics which
/// may increase or decrease in value. The behavior is to wrap around on
/// overflow and underflow.
///
/// Common examples are queue depths, temperatures, and usage metrics.
///
/// Unlike a standard `Gauge`, a `LazyGauge` will not report a value unless it
/// has been initialized by writing to at least once. This is useful for when
/// you want to declare metrics statically, but only report metrics that are
/// being used.
pub type LazyGauge = Lazy<Gauge>;

#[doc(hidden)]
pub mod export {
    pub use metriken_core::declare_metric_v1;
}
