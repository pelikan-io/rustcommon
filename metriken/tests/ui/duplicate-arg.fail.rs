#[allow(unused_imports)]
use metriken::{metric, Counter};

#[metric(
    name = "a",
    name = "b"
)]
static DUMMY: Counter = Counter::new();

fn main() {}
