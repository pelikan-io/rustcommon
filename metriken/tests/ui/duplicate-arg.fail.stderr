error: unexpected duplicate argument `name`
 --> tests/ui/duplicate-arg.fail.rs:6:5
  |
6 |     name = "b"
  |     ^^^^
